subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -237,40 +260,65 @@ public FileCollection fileCollection(Dependency... dependencies) {
         return new ConfigurationFileCollection(WrapUtil.toLinkedSet(dependencies));
     }
 
-    public void includedInResolveResult() {
-        includedInResult = true;
+    public void markAsObserved() {
+        synchronized (lock) {
+            if (state == InternalState.UNOBSERVED) {
+                state = InternalState.OBSERVED;
+            }
+        }
         for (Configuration configuration : extendsFrom) {
-            ((ConfigurationInternal) configuration).includedInResolveResult();
+            ((ConfigurationInternal) configuration).markAsObserved();
         }
     }
 
     public ResolvedConfiguration getResolvedConfiguration() {
-        resolveNow();
+        resolveNow(InternalState.RESULTS_RESOLVED);
         return cachedResolverResults.getResolvedConfiguration();
     }
 
-    private void resolveNow() {
+    private void resolveNow(InternalState requestedState) {
         synchronized (lock) {
-            if (state == State.UNRESOLVED) {
+            boolean needsResolve = state == InternalState.UNOBSERVED || state == InternalState.OBSERVED || modified;
```

Comment:
Maybe this is cleaner:

```java
private void resolveNow(InternalState requestedState) {
    synchronized (lock) {
        boolean needsResolve = false;
        switch (state) {
            case UNOBSERVED:
            case OBSERVED:
                needsResolve = true;
                break;
            case TASK_DEPENDENCIES_RESOLVED:
                needsResolve = modified && requestedState == InternalState.RESULTS_RESOLVED;
                break;
            case RESULTS_RESOLVED:
                break;
        }
        if (needsResolve) {
            DependencyResolutionListener broadcast = getDependencyResolutionBroadcast();
            ResolvableDependencies incoming = getIncoming();
            broadcast.beforeResolve(incoming);
            cachedResolverResults = resolver.resolve(this);
            for (Configuration configuration : extendsFrom) {
                ((ConfigurationInternal) configuration).markAsObserved();
            }
            resolvedWithFailures = cachedResolverResults.getResolvedConfiguration().hasError();
            markAsResolved(requestedState);
            broadcast.afterResolve(incoming);
        } else {
            if (modified) {
                DeprecationLogger.nagUserOfDeprecatedBehaviour(String.format("Attempting to resolve %s that has been resolved previously. Changes made since the configuration was originally resolved are ignored", getDisplayName()));
            }
            markAsResolved(requestedState);
        }
    }
}

private void markAsResolved(InternalState requestedState) {
    // We can't move back from resolved for results state
    if (state != InternalState.RESULTS_RESOLVED) {
        state = requestedState;
    }
    modified = false;
}
```